shamir_mnemonic/share.py

```diff
@@ -0,0 +1,158 @@
+from typing import Iterable, List, NamedTuple
+
+import attr
+
+from . import rs1024, wordlist
+from .constants import (
+    ID_EXP_LENGTH_WORDS,
+    ITERATION_EXP_LENGTH_BITS,
+    METADATA_LENGTH_WORDS,
+    MIN_MNEMONIC_LENGTH_WORDS,
+    RADIX,
+    RADIX_BITS,
+)
+from .utils import MnemonicError, bits_to_bytes, bits_to_words, int_to_indices
+
+WordIndex = int
+
+
+def _int_to_word_indices(value: int, length: int) -> List[WordIndex]:
+    """Converts an integer value to a list of base 1024 indices in big endian order."""
+    return list(int_to_indices(value, length, radix_bits=RADIX_BITS))
+
+
+def _int_from_word_indices(indices: Iterable[WordIndex]) -> int:
+    """Converts a list of base 1024 indices in big endian order to an integer value."""
+    value = 0
+    for index in indices:
+        value = value * RADIX + index
+    return value
+
+
+class ShareSetParameters(NamedTuple):
+    identifier: int
+    iteration_exponent: int
+    group_threshold: int
+    group_count: int
+
+
+@attr.s(auto_attribs=True, frozen=True)
+class Share:
+    """Represents a single mnemonic share and its metadata"""
+
+    identifier: int
+    iteration_exponent: int
+    group_index: int
+    group_threshold: int
+    group_count: int
+    index: int
+    threshold: int
+    value: bytes
+
+    def common_parameters(self) -> ShareSetParameters:
+        """Returns values that uniquely identify a matching set of shares."""
+        return ShareSetParameters(
+            self.identifier,
+            self.iteration_exponent,
+            self.group_threshold,
+            self.group_count,
+        )
+
+    def _encode_id_exp(self) -> List[WordIndex]:
+        id_exp_int = (
+            self.identifier << ITERATION_EXP_LENGTH_BITS
+        ) + self.iteration_exponent
+        return _int_to_word_indices(id_exp_int, ID_EXP_LENGTH_WORDS)
+
+    def _encode_group_params(self) -> List[WordIndex]:
```

Comment:
doesn't "metadata" include the checksum though?

Reply:
You are right, the constant `METADATA_LENGTH_WORDS` includes the length of the checksum. So I would either call it `_encode_share_params()` or replace it with a function called `_encode_share_header()` which would return the first four indices including id and exp.

Reply:
changed to `_encode_share_params`; i think the reason I kept this as two functions is that each encodes a separate "number"

the use of `all_group_params` in `_decode_mnemonics` is OK, i presume? since these are actually the group params that are not share-specific.